src/connection.rs

```diff
@@ -1,41 +1,360 @@
-use std::ffi::c_void;
+use std::slice;
+use std::{ffi::c_void, ptr::null};
 
 use libc::{size_t, EIO};
-use rustls::Session;
+use rustls::{Certificate, ClientSession, ServerSession, Session, SupportedCipherSuite};
 
-use crate::error::rustls_io_error;
-use crate::io::{CallbackReader, CallbackWriter, ReadCallback, WriteCallback};
+use crate::{
+    cipher::{rustls_certificate, rustls_supported_ciphersuite},
+    error::{map_error, rustls_io_error, rustls_result},
+    io::{rustls_read_callback, rustls_write_callback},
+    try_callback, try_mut_slice,
+};
+use crate::{ffi_panic_boundary, try_ref_from_ptr};
+use crate::{
+    io::{CallbackReader, CallbackWriter, ReadCallback, WriteCallback},
+    is_close_notify,
+};
+use crate::{try_mut_from_ptr, try_slice, userdata_push, CastPtr};
+use rustls_result::NullParameter;
 
-// Call Session::read_tls, providing an &mut dyn Write implemented with a C callback.
-pub(crate) fn read_tls(
-    session: &mut dyn Session,
-    callback: ReadCallback,
+pub(crate) struct Conn {
+    pub(crate) conn: Inner,
+    pub(crate) userdata: *mut c_void,
+}
+
+pub(crate) enum Inner {
```

Comment:
nit: I think it'd be nicer if the `conn` member were more completely abstracted away from clients. At a glance, there is one spot where a `struct Conn[ection]` needs to be handled as a concrete `ServerSession` (`rustls_server_session_get_sni_hostname`) and otherwise clients are expected to stick to methods in `trait Session`. I think the `enum Inner` type and the `conn` member on `struct Conn[ection]` should be private to `connection.rs`, and you can enable `rustls_server_session_get_sni_hostname` by providing something like a `TryFrom` that attempts to convert `struct Conn` into `ServerSession`. However the whole thing is private to the crate so I don't think it's a big deal.

Reply:
Ah this is a good idea. Can't do `TryFrom` since we're working with refs, not owned values. But I think `AsRef<ServerSession>` and so on can work.